File: 507.perfect-number.cpp
/*
 * @lc app=leetcode id=507 lang=cpp
 *
 * [507] Perfect Number
 */

#include <bits/stdc++.h>
#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  bool checkPerfectNumber(int num) {
    if (num <= 1) {
      return false;
    }
    int sum = 1;
    int max = sqrt(num);
    for (int i = 2; i <= max; i++) {
      if (num % i == 0) {
        sum += i + (num / i);
      }
    }
    return sum == num;
  }
};
// @lc code=end

int main() {
  Solution s;
  assert(s.checkPerfectNumber(6));
  assert(s.checkPerfectNumber(28));
  assert(s.checkPerfectNumber(496));
  assert(s.checkPerfectNumber(8128));
  assert(!s.checkPerfectNumber(2));
  assert(!s.checkPerfectNumber(100000000));
  assert(!s.checkPerfectNumber(1));
  return 0;
}

File: 507.perfect-number.rs
/*
 * @lc app=leetcode id=507 lang=rust
 *
 * [507] Perfect Number
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn check_perfect_number(num: i32) -> bool {
        if num <= 1 {
            return false;
        }
        let mut sum = 1;
        let max = (num as f64).sqrt() as i32;
        for i in 2..=max {
            if num % i == 0 {
                sum += i + (num / i);
            }
        }
        sum == num
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::check_perfect_number(28));
}

#[test]
fn test2() {
    assert!(!Solution::check_perfect_number(7));
}
